### ISSUE TITLE
PDF size greatly increases when resaving.

### ISSUE THREAD

**Issue by BBBalls** [podofo/podofo#163]
* **I'm submitting a:**
  - [ ] Vulnerability report -> **use the specific [form](https://github.com/podofo/podofo/security/advisories/new) instead!**
  - [ x] Bug report
  - [ ] Feature request/suggestion

* **[Bug report] What is the current behavior?**

I encounter this bug while using [Calibre](https://calibre-ebook.com/) to embed some metadata into a couple of PDFs. I reported the bug to the Calibre bug tracker, [Bug #2066529](https://bugs.launchpad.net/calibre/+bug/2066529). The Calibre dev said the issue needs to be reported here.

When applying metadata to the PDFs the go from less than 10MB to more than 100MB. There does seem to be something peculiar with the PDFs, because this issue does not occur when apply metadata to other PDFs.

The bug occurs with the two PDFs that can be found on this page: https://www.buddhism.hku.hk/publication-post/the-path-to-freedom-vimuttimagga/

The title of the two volume book is *The Path to Freedom Vimuttimagga*.

The Calibre dev says simply opening and saving the PDF (presumably with podofo via the Calibre UI) inflates the size of the PDFs.

* **[Bug report] What is the expected behavior?**

The expected behavior is for the PDFs to not grow significantly in size.

* **[Bug report] Please provide the steps to reproduce and if possible a minimal reproduction code of the problem**

I am not at all familiar with you project, so I can only describe how to reproduce the bug via Calibre.

- Import one the PDFs from the link about into a Calibre Library.
- Open the preferences menu.
  - Select 'Toolbars and menus'
  - Select 'The main toolbar' in the dropdown menu
  - Add 'Embed metadata' from the 'Available actions' column to the 'Current Actions' column.
  - Apply the changes and restart Calibre.
- Select the PDF
- Press the 'Embed metadata' button on the main menu bar.
- After the embedding metadata function is run the file size of the PDF is greatly inflated.

The Calibre dev said this command might show something meaningful `calibre-debug -c 'from calibre.utils.podofo import *; test_roundtrip("t.pdf", "x.pdf")'`

* **Please tell us about your environment:**

I am not at all familiar of the build environment in which your project is implemented in Calibre.

I have encountered this bug on Linux POP!_OS 22.04 with Calibre 7.10 and 7.11 installed via the supported Linux installation script. I  assume podofo is statically linked within the Calibre package.

**Comment by ceztko** [comment]
I referenced this issue by mistake in a commit (that I removed immediately after). Anyway: I tested and I can reproduce (both 0.10.x and master) with the following lines:

```
PdfMemDocument document;
document.Load(R"(D:\Downloads\The-Path-to-Freedom.pdf)");
document.GetMetadata().SetAuthor(PdfString("Hello"));
document.Save(R"(D:\test.pdf)");
```

For me saving was very slow, you didn't mention if it was slow for you as well. Unfortunately I have my priorities on this project and I can't guarantee I will look at this issue very soon, even though I understand it looks quite a serious one. It would help me if you could tell me how Calibre behaves in a version before 6.18, because I believe it was the first to port to PoDoFo 0.10.x, which is the new course for the library. Before it used 0.9.x which is the legacy library which I don't use/compile anymore, but it would be great to know if this issue is a regression of 0.10.x or not. There have been some regressions during the transition to 0.10.x and @kovidgoyal reported few issues, but all of them were solved.

**Comment by kovidgoyal** [comment]
This is the behavior of podofo 0.9 
```
time calibre-debug -c 'from calibre.utils.podofo import *; p = get_podofo().PDFDoc(); p.open("t.pdf"); p.save("x.pdf");'; ls -sh t.pdf x.pdf
WARNING: There are more objects (56874) in this XRef table than specified in the size key of the trailer directory (56822)!

real	6.914
user	6.231
sys	0.543
maxmem	1155 MB
faults	0
7.2M t.pdf  103M x.pdf
```

Takes seven seconds and balloons the size. Podofo 0.10:

```
time calibre-debug -c 'from calibre.utils.podofo import *; p = get_podofo().PDFDoc(); p.open("t.pdf"); p.save("x.pdf");'; ls -sh t.pdf x.pdf         

real	4.628
user	4.076
sys	0.521
maxmem	1440 MB
faults	0
7.2M t.pdf  104M x.pdf
```

Takes 5 seconds and balloons the size, without a warning about the XRef table.

**Comment by BBBalls** [comment]
I do experience a significant lag when applying metadata to the specific PDFs. The desktop environment will offer to kill the process or wait.

If you need some information beyond what kovidgoyal provided, I can install Calibre 5.37 from the distribution repo.

**Comment by ceztko** [comment]
I analyzed the situation: the good news is that there are no regressions compared to 0.9.x, the bad news is that this is a big feature missing in PoDoFo, and not really a bug. Basically the document has a massive element called a "Structure hierarchy" (`/StructTreeRoot 2477 0 R`) which describes the logical construction and semantics of the document rather than raw rendering directives. This element is not required at all for correct rendering of the document, but it may supply some easy to parse information for secondary processing. In some documents this is huge and in fact Adobe introduced a a way to compress such big elements, called _Object streams_ ("7.5.7 Object streams" in ISO 32000-2:2020), which are basically lists of indirect objects that can be further compressed by mean of regular PDF filters (typically `/FlateDecode`). The issue with PoDoFo is that currently it can't preserve this optimized storage, and it will end to unroll all these objects and store them inefficiently in separated uncompressed objects. I will have to evaluate how hard is to preserve these object streams, which I currently don't know (it may be quite a big task).

Before that the only possible (not entirely safe) workaround is deleting /StructTreeRoot from the catalog, which is unnecessary for rendering, but may be required in some other context.

**Comment by ceztko** [comment]
I investigated the issue a little bit further: the problem is basically the parser eagerly loading all compressed objects, forgetting the reference to the containing object streams. This removes the chance for the saving procedure to preserve unmodified objects in their containing stream. If we inherit `PdfObject` to specialize for such compressed objects, we could treat them differently, lazily decompressing the parent streams and invalidating them only if at least one of the compressed objects in the stream was modified. This modification requires some good knowledge of the PDF specification but it shouldn't be extremely complex. I will do it at some point but I can't guarantee when I will start working on it as I have other priorities for now.

**Comment by ceztko** [comment]
This was a big one. Some acknowledgements: AI assisted (Opus 5) it took a couple of days. Before AI it would have taken at least 10 days of total immersion, and test coverage would have been lacking compared to what was done. Many unrelated bugs were also discovered, which in pre-AI times would have been missed or the fix delayed. Reading back the code helped with small oversights and code quality assurance. The biggest mistake it did was trying to add a public API flag that could have corrupted files in some cases. It resolved easily by convincing the assistant to solve it through private/internal methods instead.

@kovidgoyal Can you make sure Calibre compiles with modern PoDoFo? I read some unconfirmed reports that were saying that Calibre is not compatible with PoDoFo >=1.0 . I looked at your code and you were not using too much stuff, and I'm sure you'll find just small issues (if any). Starting with 1.0 API is stable, with changes allowed only in limited parts of the API that are not widely used and/or very new.

**Comment by kovidgoyal** [comment]
Yes it doesnt. There were some blockers IIRC for instance it uses PdfIndirectObjectList::RemoveObject() which has become private.

**Comment by kovidgoyal** [comment]
Also it doesnt seem to build on windows without fontconfig anymore.  I guess I can disable font manager or use GDI, no big deal.

**Comment by kovidgoyal** [comment]
Now ported, I worked around missing RemoveObject() by replacing the objects with a null object which removes most of the object just leaving an XREF to it: https://github.com/kovidgoyal/calibre/commit/dce04dbaddb3f4d4606c1299976d1268ab7275a3

**Comment by ceztko** [comment]
> it uses PdfIndirectObjectList::RemoveObject() which has become private.

Now garbage collection is performed on `Save()`, so you can just remove the object from the document tree and it will be removed from list of indirect objects automatically.

> Also it doesnt seem to build on windows without fontconfig anymore. I guess **I can disable font manager or use GDI**, no big deal.

That's right. Now, using FontManager with FontConfig is highly recommended, but you can still distribute the software without font lookup or with the old GDI fallback system if you prefer.
